modules/imgcodecs/src/loadsave.cpp

```diff
@@ -257,6 +257,46 @@ static ImageDecoder findDecoder( const String& filename ) {
     return ImageDecoder();
 }
 
+static ImageDecoder findDecoder( const String& filename, ImreadError& error ) {
+
+    size_t i, maxlen = 0;
+
+    /// iterate through list of registered codecs
+    ImageCodecInitializer& codecs = getCodecs();
+    for( i = 0; i < codecs.decoders.size(); i++ )
+    {
+        size_t len = codecs.decoders[i]->signatureLength();
+        maxlen = std::max(maxlen, len);
+    }
+
+    /// Open the file
+    FILE* f= fopen( filename.c_str(), "rb" );
+
+    /// in the event of a failure, return an empty image decoder
+    if( !f ) {
+        error = IMREAD_ERROR_FILE_NOT_FOUND;
+        CV_LOG_WARNING(NULL, "imread_('" << filename << "'): can't open/read file: check file path/integrity");
+        return ImageDecoder();
+    }
+
+    // read the file signature
+    String signature(maxlen, ' ');
+    maxlen = fread( (void*)signature.c_str(), 1, maxlen, f );
```

Comment:
c_str() returns `const char*` for a reason. Casting and writing to that pointer should be avoided (it could be a shadow copy).

Mutable `std::vector<uint8_t>` or `cv::AutoBuffer<int8_t>` should be used instead (perhaps including `.checkSignature()`)